### PR TITLE
DP-527 - Prevent empty VAT from being persisted

### PR DESCRIPTION
**Pull request by dbgoaco**
[DP-527](https://noticingsystem.atlassian.net/jira/software/projects/DP/boards/22?selectedIssue=DP-527)

### CODE REVIEW

**Review by dharmverma — APPROVED** [review]

File: Services/CO.CDP.Organisation.WebApi.Tests/UseCase/UpdateOrganisationUseCaseTest.cs [hunk 1]
```diff
@@ -202,6 +202,40 @@ public async Task Execute_ShouldPublishOrganisationUpdatedEvent_WhenOrganisation
...
+        var command = new UpdateOrganisation
+        {
+            Type = OrganisationUpdateType.AdditionalIdentifiers,
+            Organisation = new OrganisationInfo
+            {
+                AdditionalIdentifiers = [new OrganisationIdentifier
+                {
+                    Id = "",
+                    LegalName = "Acme",
+                    Scheme = "GB-VAT-EMPTY"
```
Comment (jakzal) [inline Services/CO.CDP.Organisation.WebApi.Tests/UseCase/UpdateOrganisationUseCaseTest.cs:217]: This isn’t  ascheme we expect. It’s just “VAT” I think.

Comment (dbgoaco) [reply Services/CO.CDP.Organisation.WebApi.Tests/UseCase/UpdateOrganisationUseCaseTest.cs]: Done.

File: Services/CO.CDP.Organisation.WebApi.Tests/UseCase/UpdateOrganisationUseCaseTest.cs [hunk 2]
```diff
@@ -202,6 +202,39 @@ public async Task Execute_ShouldPublishOrganisationUpdatedEvent_WhenOrganisation
...
+        _organisationRepositoryMock.Setup(repo => repo.Find(_organisationId)).ReturnsAsync(organisation);
+
+        var result = await UseCase.Execute((_organisationId, command));
+
+        result.Should().BeTrue();
+        _organisationRepositoryMock.Verify(repo => repo.Save(organisation!), Times.Once);
+
+        organisation.Identifiers.FirstOrDefault(i => i.Scheme == "GB-VAT").Should().BeNull();
+    }
+
```
Comment (jakzal) [inline Services/CO.CDP.Organisation.WebApi.Tests/UseCase/UpdateOrganisationUseCaseTest.cs:237]: This only handles the case of adding an identifier. What about the case when someone adds VAT but then changes their answer?

File: Services/CO.CDP.Organisation.WebApi.Tests/UseCase/UpdateOrganisationUseCaseTest.cs (hunk 1 above)
Comment (dbgoaco) [reply Services/CO.CDP.Organisation.WebApi.Tests/UseCase/UpdateOrganisationUseCaseTest.cs]: Done.

File: Services/CO.CDP.Organisation.WebApi.Tests/UseCase/UpdateOrganisationUseCaseTest.cs (hunk 2 above)
Comment (jakzal) [reply Services/CO.CDP.Organisation.WebApi.Tests/UseCase/UpdateOrganisationUseCaseTest.cs:237]: I mean the situation when someone gives the VAT number at first:
image

But then, they change their answer to "No, I don't have a VAT number":

image

**Review by jakzal — CHANGES_REQUESTED** [review]

File: Services/CO.CDP.Organisation.WebApi.Tests/UseCase/UpdateOrganisationUseCaseTest.cs (hunk 2 above)
Comment (dbgoaco) [reply Services/CO.CDP.Organisation.WebApi.Tests/UseCase/UpdateOrganisationUseCaseTest.cs:265]: Added Execute_ShouldUpdateVatNumber_WhenVatIdentifierAlreadyExists

**Comment by jakzal** [comment]
Now, if I select "No", the answer is not persisted as the supplier information screen doesn't know I answered the question:

image

It seems that `completed_vat` flag is not set in `supplier_information` table.

File: Services/CO.CDP.Organisation.WebApi/UseCase/UpdateOrganisationUseCase.cs [hunk 3]
```diff
@@ -116,4 +123,35 @@ public async Task<bool> Execute((Guid organisationId, UpdateOrganisation updateO
...
+        {
+            nextPrimaryIdentifier = organisation.Identifiers.FirstOrDefault(i =>
+                i.Scheme == AssignIdentifierUseCase.IdentifierSchemes.Ppon);
+
+            if (nextPrimaryIdentifier == null)
+            {
+                nextPrimaryIdentifier = organisation.Identifiers.FirstOrDefault(i =>
+                    i.Scheme == AssignIdentifierUseCase.IdentifierSchemes.Other);
+            }
+        }
```
Comment (jakzal) [inline Services/CO.CDP.Organisation.WebApi/UseCase/UpdateOrganisationUseCase.cs:146]: I'm not sure about this code. It only seems to be looking at PPON and Other type of identifiers. Doesn't seem to be right.

Comment (dbgoaco) [reply Services/CO.CDP.Organisation.WebApi/UseCase/UpdateOrganisationUseCase.cs]: Now checks whether the primary identifier is being removed. If it is, it finds the next primary identifier, if there is one.

**Review by jakzal — APPROVED** [review]

**Pull request merged**